llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp

```diff
@@ -439,6 +444,264 @@ struct PromoteMem2Reg {
   }
 };
 
+/// This class computes liveness and dominance frontier (DF) of allocas in
+/// batch. The scalar Mem2Reg algorithm processes alloca one by one, which is
+/// not efficient in general because it repeatedly traverses the CFG and
+/// enumerates predecessors or successors of basic blocks. Given that most
+/// blocks are likely covered by the liveness of more than one values,
+/// processing these values at one go can reduce the amount of CFG traversal.
+/// This requires certain modifications to the algorithm computing liveness
+/// range and DF of allocas. In general, previously we use a set to track
+/// visited blocks, which is equivalent to have each block keeping track of a
+/// "visited" state, and now the visited state becomes a bit vector
+/// corresponding to each alloca to be process in one go. We OR the states when
+/// visiting a block and if the state is changed we need to keep traversing
+/// (because info carried by one alloca from a certain edge has not converged
+/// yet) again, otherwise we have reached a fixed point and can stop there. The
+/// number of repeated traversal is bounded by the bit vector's width because
+/// each time at least a bit is set, so the overall complexity will be less than
+/// the sum of each alloca being processed individually.
+class VectorizedMem2Reg {
```

Comment:
I am considering to generalize it if there is another application of this algorithm. There are several places that need to traverse the CFG for every variable, but I have not identified any being a bottleneck yet.

Reply:
Have a vectorized IDF will probably make the API simpler and make the rest of the change more readable. It will also lower the barrier for future adoption.